File: java_CI_scripts/src/collection.clj
(ns collection
  "Handles the bioimageio collection of rdfs/models"
  (:require [config :refer [ROOTS]]
            [babashka.fs :as fs]
            [cheshire.core :as json]))

(defn str-json->vector
  "Returns the parsed list of resources/versions to test, given a raw json string"
  [str-json]
  (let [parsed (json/parse-string str-json true)]
    (:include parsed)))

(defn file-json->vector
  "Returns the parsed list of resources/versions to test, given a json file"
  [file-json]
  (str-json->vector (slurp file-json)))

(defn filter-rdfs
  "Keeps only the rdf.yaml from a seq of paths"
  [paths]
  (filter #(= (fs/file-name %) "rdf.yaml") paths))

(defn resource->paths
  "Takes a resource/version to test and returns list of rdf path(s) (multiple if globbing).
  If only 1 argument is given, uses COLLECTION-ROOT as root path"
  ([root resource-map]
   (let [{:keys [resource_id version_id]} resource-map]
     (cond
       (= resource_id "**") (filter-rdfs (fs/glob root "**"))
       (= version_id "**") (filter-rdfs (fs/glob (fs/path root resource_id) "**"))
       :else (vector (fs/path root resource_id version_id "rdf.yaml")))))
  ([resource-map] (resource->paths (:collection-root ROOTS) resource-map)))

(defn get-rdfs-to-test
  "Compiles a list of rdf paths that need to be tested, given a list of resource/versions maps.
  If only 1 argument is given, uses COLLECTION-ROOT as root path"
  ([root resources-vector] (set (flatten (map #(resource->paths root %) resources-vector))))
  ([resources-vector] (get-rdfs-to-test (:collection-root ROOTS) resources-vector)))

(defn parse-collection
  "Parse the contents of collection.json into a hash-map"
  ([] (parse-collection (fs/file (:collection-root ROOTS) ".." "collection.json")))
  ([collection-file]
   (-> (fs/file collection-file)
       slurp
       (json/parse-string true)
       :collection)))

(defn get-model-identifier
  "Gets the resource-id and the latest version-id of a collection model (in collection.json)"
  [collection-model]
  {:resource_id (:id collection-model) :version_id (first (:versions collection-model))})

(defn generate-pending-matrix-from-collection
  "Generates the pending matrix corresponding to the models in the collection.json "
  ([verbose?] (generate-pending-matrix-from-collection
                (fs/file (:pending-matrix-root ROOTS) "only-collection.json") verbose?))
  ([output-file verbose?]
   (let [models (filter #(= (:type %) "model") (parse-collection))
         pending-matrix (map get-model-identifier models)]
     (spit output-file (json/generate-string {:include pending-matrix} {:pretty true}))
     (if verbose?
       (printf "Pending matrix with %3d entries to test created in %s\n" (count models) output-file)
       (flush))
     (count models))))

File: python_CI_scripts/collection.py
"""Handles the bioimageio collection of rdf's"""

from config import ROOTS
from pathlib import Path
import json
import itertools


def str_json_2_vector(str_json):
    """Returns the parsed list of resources/versions to test, given a raw json string"""
    try:
        parsed = json.loads(str_json)
        return parsed["include"]
    except json.JSONDecodeError as e:
        print("String is not a valid json")
        return []


def file_json_2_vector(json_file):
    """Returns the parsed list of resources/versions to test, given a json file"""
    with open(json_file, "r") as file:
        data = file.read().rstrip()
        return str_json_2_vector(data)


def filter_rdfs(paths):
    """Keeps only the rdf.yaml from a seq of paths"""
    return list(filter(lambda x: x.name == "rdf.yaml", paths))


def resources_2_paths(resource_map, root=ROOTS["collection-root"]):
    """Takes a resource/version to test and returns list of rdf path(s) (multiple if globbing).
    If only 1 argument is given, uses COLLECTION-ROOT as root path"""
    r_id = resource_map["resource_id"]
    v_id = resource_map["version_id"]

    if r_id == "**":
        return filter_rdfs(root.glob("**/*"))
    elif v_id == "**":
        return filter_rdfs((root / r_id).glob("**/*"))
    else:
        return [Path(root / r_id / v_id / "rdf.yaml")]


def get_rdfs_to_test(resources_vector, root=ROOTS["collection-root"]):
    """ Compiles a list of rdf paths that need to be tested, given a list of resource/versions maps.
    If only 1 argument is given, uses COLLECTION ROOT as root path"""
    unflattened = list(map(lambda x: resources_2_paths(x, root), resources_vector))
    return list(set(itertools.chain(*unflattened)))
